here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/models/storage/HeapCacheCursor.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2017-2023 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.lib.core.models.storage;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * HeapCacheCursor reads all data (up to provided limit) ahead of time to memory, and that allows to go back and forth.
+ * It's not thread-safe in any way, {@link SeekableCursor} is designed to jump to any position of the cursor, so it doesn't really make sense to use it in parallel.
+ * Use {@link HeapCacheCursor} only when you need to walk through results multiple times, if you need to read data only once - please use {@link ForwardCursor} as it consumes much less memory.
+ *
+ * @param <FEATURE>
+ * @param <CODEC>
+ */
+public class HeapCacheCursor<FEATURE, CODEC extends FeatureCodec<FEATURE, CODEC>>
+    extends SeekableCursor<FEATURE, CODEC> {
+
+  protected static final long BEFORE_FIRST_POSITION = -1;
+
+  protected final ForwardCursor<FEATURE, CODEC> originalCursor;
+
+  protected final Map<Long, ForwardCursor<FEATURE, CODEC>.Row> inMemoryData = new HashMap<>();
+
+  protected final long limit;
+  protected final boolean reOrder;
+  protected long positionOfLastElement = BEFORE_FIRST_POSITION;
+
+  public HeapCacheCursor(
+      @NotNull FeatureCodecFactory<FEATURE, CODEC> codecFactory,
+      long limit,
+      boolean reOrder,
+      @NotNull ForwardCursor<FEATURE, CODEC> originalCursor) {
+    super(codecFactory);
+    this.originalCursor = originalCursor;
+    this.position = originalCursor.position;
+    this.limit = limit;
+    // TODO FIXME re-order elements if needed
+    this.reOrder = reOrder;
+
+    final boolean isReadAllRequested = limit == -1;
+
+    long count = 0;
+    while (originalCursor.hasNext() && (isReadAllRequested || count < limit)) {
```

Comment:
It reads all data ahead of time, I can make it lazy loading, but I don't know if there is a point to do that